swarm/src/handler/one_shot.rs

```diff
@@ -213,7 +213,8 @@ where
             }
             ConnectionEvent::DialUpgradeError(DialUpgradeError { error, .. }) => {
                 if self.pending_error.is_none() {
-                    self.pending_error = Some(error);
+                    log::debug!("DialUpgradeError: {error}");
```

Comment:
I'd like to use `crate::print_error_chain` here and not within the `Display` impl please.

Reply:
I am afraid it is not that simple. 

- `print_error_chain` requires a `Formatter` and is invoked within implementations of the `Display` trait. So I can't use it directly here
```rust
fn print_error_chain(f: &mut fmt::Formatter<'_>, e: &dyn error::Error) -> fmt::Result {...}
```

- the type of `error` contained in `DialUpgradeError` is `ConnectionHandlerUpgrErr<InnerError>`, where `InnerError` is not bounded to `error::Error`, thus I had to introduce this bound in the `Display` impl for `ConnectionHandlerUpgrErr`.

Reply:
Right, I forgot about that, sorry for the confusion!

I think it is okay to continue using `print_error_chain` here if we consequently don't return the inner error from `source`.

If we do this, I'd like to see a unit test that ensures the `Error` impl is not changed.

Reply:
Alternatively, we can create a local one-off use newtype here that uses `print_error_chain` in its display function. We can then construct this type and immediately use it to print the error.

Reply:
I believe it is reasonable to [mimic](https://github.com/libp2p/rust-libp2p/blob/69af8506d1eb17f5d11b7364693d45b322205180/swarm/src/lib.rs#L1747) `impl fmt::Display for DialError`
- `DialEerror` is _structurally_ very similar to `ConnectionHandlerUpgrErr`,
- `DialError` already sets an example of how to use `print_error_chain`.

It feels right that the `Display` knows how to print the error info correctly, thus my proposition would be:

```rust
impl<TUpgrErr> fmt::Display for ConnectionHandlerUpgrErr<TUpgrErr>
where
    TUpgrErr: error::Error + 'static,
{
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            ConnectionHandlerUpgrErr::Timeout => {
                write!(f, "Timeout error while opening a substream")
            }
            ConnectionHandlerUpgrErr::Timer => {
                write!(f, "Timer error while opening a substream")
            }
            ConnectionHandlerUpgrErr::Upgrade(err) => {
                write!(f, "DialUpgradeErr: ")?;
                print_error_chain(f, err)
            }
        }
    }
}
```

Otherwise, let's just use `log::debug!("DialUpgradeError: {error}");`

Reply:
Your proposal sounds good! I think we should also return `None` in `source` then, otherwise we might get double printing.

Reply:
Struggling to understand where you'd like to return `None`. The `error::Error` implementation of `` already returns `None` where appropriate
```rust
impl<TUpgrErr> error::Error for ConnectionHandlerUpgrErr<TUpgrErr>
where
    TUpgrErr: error::Error + 'static,
{
    fn source(&self) -> Option<&(dyn error::Error + 'static)> {
        match self {
            ConnectionHandlerUpgrErr::Timeout => None,
            ConnectionHandlerUpgrErr::Timer => None,
            ConnectionHandlerUpgrErr::Upgrade(err) => Some(err),
        }
    }
}
```
Am I missing something?

Reply:
If we print the sources as part of `Display::fmt`, we should return `None` on within `source`, otherwise libraries like `anyhow` which walk down the source chain will print all sources again, leading to duplicate error messages for the user.